python/cudf/cudf/core/dataframe.py

```diff
@@ -907,14 +907,24 @@ def _init_from_dict_like(
         if index is None:
             num_rows = 0
             if data:
-                col_name = next(iter(data))
-                if is_scalar(data[col_name]):
-                    num_rows = num_rows or 1
+                lengths = set()
+                for key, value in data.items():
+                    if is_scalar(value):
+                        data[key] = value
+                        lengths.add(1)
+                    else:
+                        value = column.as_column(
+                            value, nan_as_null=nan_as_null
+                        )
+                        lengths.add(len(value))
+                        data[key] = value
+
+                if len(lengths - {1}) > 1:
+                    raise ValueError("All arrays must be of the same length")
                 else:
-                    data[col_name] = column.as_column(
-                        data[col_name], nan_as_null=nan_as_null
+                    num_rows = (
+                        1 if lengths == {1} else next(iter(lengths - {1}))
                     )
-                    num_rows = len(data[col_name])
```

Comment:
Is this rewrite too tricksy?
```python
            keys, values, lengths = zip(
                *(
                    (k, v, 1)
                    if is_scalar(v)
                    else (k, vc := as_column(v, nan_as_null=nan_as_null), len(vc))
                    for k, v in data.items()
                )
            )
            data = dict(zip(keys, values))
            try:
                (num_rows,) = (lengths - {1}) or {1}
            except ValueError:
                raise ValueError("All arrays must be the same length")
```

I tend to write things like this because I (personally) find stateful modification in loops hard to reason about. But that is probably just me.

Actually you don't need to carry the keys along for the ride, but I tend to because I still don't trust dictionary iteration order :).

Reply:
This approach looks good to me, updated 👍